### PR TITLE
refactor(experimental): add prefixCodec to @solana/codecs-core

### PR DESCRIPTION
**Pull request by lorisleiva**
This PR adds a new `prefixCodec` primitive to `@solana/codecs-core`. This primitives accepts any codec and prefixes its size using a `NumberCodec` provided as a second argument.

For instance:

```ts
const codec = prefixCodec(getBase58Codec(), getU32Codec());

codec.encode('hello world');
// 0x0b00000068656c6c6f20776f726c64
//   |       └-- Our encoded base-58 string.
//   └-- Our encoded u32 size prefix.
```

This PR also updates the `getStringCodec` and `getBytesCodec` as they can now use the `prefixCodec` primitive to simplify their internal logic.

P.S.: Now that `getStringCodec` and `getBytesCodec` are simple aggregator of `fixCodec` and `prefixCodec`, we should consider removing them or adjusting them further. (Pros: no default values => better tree-shakeability, Cons: slightly less friendly API?).

```ts
// With getStringCodec.
getStringCodec();
getStringCodec({ encoding: getBase58Codec() });
getStringCodec({ size: 32 });
getStringCodec({ size: getU16Codec() });
getStringCodec({ size: "variable" });

// Without getStringCodec.
prefixCodec(getUtf8Codec(), getU32Codec());
prefixCodec(getBase58Codec(), getU32Codec());
fixCodec(getUtf8Codec(), 32);
prefixCodec(getUtf8Codec(), getU16Codec());
getUtf8Codec();
```

### CODE REVIEW

File: packages/codecs-core/README.md [hunk 1]
```diff
@@ -362,6 +362,31 @@ const get32BytesBase58Decoder = () => fixDecoder(getBase58Decoder(), 32);
 const get32BytesBase58Codec = () => combineCodec(get32BytesBase58Encoder(), get32BytesBase58Codec());
 ```
 
+## Prefixing the size of codecs
+
+The `prefixCodec` function allows you to store the byte size of any codec as a number prefix. This allows you to contain variable-size codecs to their actual size.
```
Comment (steveluscher) [inline packages/codecs-core/README.md:367]: I wonder if there's a name we can give this codec to make it more clear that the prefix it produces is _particularly_ the size of the field.

I wonder what other libraries call this.

* Length-delimited records ([protobuf](https://protobuf.dev/programming-guides/encoding/#length-types))
* Length-prefix framing

Doesn't super matter. I can't find strong evidence that there's even a consensus on what to call such a thing.

File: packages/codecs-core/README.md [hunk 2]
```diff
@@ -362,6 +362,31 @@ const get32BytesBase58Decoder = () => fixDecoder(getBase58Decoder(), 32);
...
+//   |       └-- Our encoded base-58 string.
+//   └-- Our encoded u32 size prefix.
+```
+
+You may also use the `prefixEncoder` and `prefixDecoder` functions to separate your codec logic like so:
+
+```ts
+const getU32Base58Encoder = () => prefixEncoder(getBase58Encoder(), getU32Codec());
+const getU32Base58Decoder = () => prefixDecoder(getBase58Decoder(), getU32Codec());
+const getU32Base58Codec = () => combineCodec(getU32Base58Encoder(), getU32Base58Codec());
```
Comment (steveluscher) [inline packages/codecs-core/README.md:387]: This looks like it needs another pass.

**Review by steveluscher — CHANGES_REQUESTED** [review]
tl;dr unless I missed it, I _think_ you need something to prevent the actual size of the data from being larger than that which you can represent using the length prefix. We either need to throw when that's the case, or truncate the data to the actual length that will eventually be decoded.

File: packages/codecs-core/src/prefix-codec.ts [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    prefix: FixedSizeNumberEncoder,
+): FixedSizeEncoder<TFrom>;
+export function prefixEncoder<TFrom>(encoder: Encoder<TFrom>, prefix: NumberEncoder): VariableSizeEncoder<TFrom>;
+export function prefixEncoder<TFrom>(encoder: Encoder<TFrom>, prefix: NumberEncoder): Encoder<TFrom> {
+    const write = ((value, bytes, offset) => {
+        // Here we exceptionally use the `encode` function instead of the `write`
+        // function to contain the content of the encoder within its own bounds.
+        const encoderBytes = encoder.encode(value);
+        offset = prefix.write(encoderBytes.length, bytes, offset);
+        bytes.set(encoderBytes, offset);
```
Comment (steveluscher) [inline packages/codecs-core/src/prefix-codec.ts:46]: We have to do _something_ to prevent this scenario, right?

```ts
const { decode, encode } = prefixCodec(getUtf8Encoder(), getU8Encoder());
const bytes = encode('string longer [...] than 256 characters');
bytes >= 257; // true
const str = decode(bytes); // string longer [...] than 256 ch
// And now the buffer after that point is corrupt
```

Comment (lorisleiva) [reply packages/codecs-core/src/prefix-codec.ts:46]: I'll had a test to be explicit about it but if the u8 codec isn't able to store the string size, it will throw an error.

It's the responsibility of the size codec to determine if it can indeed encode/decode the provided size because you could technically provide more complicated number codecs such as the `shortU16`.

Comment (steveluscher) [reply packages/codecs-core/src/prefix-codec.ts:46]: Ah, excellent.

**Review by steveluscher — APPROVED** [review]
Added via Giphy

File: packages/codecs-core/README.md (hunk 1 above)
Comment (lorisleiva) [reply packages/codecs-core/README.md:367]: Yeah naming this one has been tricky. The constraints for naming this were:

- Needs to start with a verb and finish with `Codec` since it's a codec primitive in the likes of `mapCodec` or `fixCodec`.
- Somewhat echo the `fixCodec` primitive which makes any codec a fixed-size one by providing the byte size explicitly. This is because technically, both of these codecs are responsible for "sizing" the codec they wrap.

`prefixCodec` seemed to do the job okay but I agree it's not ideal. My main issues with it are:

- Not clear that it's prefixing the size.
- A bit too close to the new `getHiddenPrefixCodec` helper at the end of this stack.

An alternative could be `prefixCodecSize` and we could rename `fixCodec` to `fixCodecSize` to bring the two closer together.

Wdyt?

Comment (lorisleiva) [reply packages/codecs-core/README.md:367]: I made the suggested change in a subsequent PR: https://github.com/solana-labs/solana-web3.js/pull/2411.

File: packages/codecs-core/src/prefix-codec.ts (hunk 3 above)
Comment (lorisleiva) [reply packages/codecs-core/src/prefix-codec.ts:46]: I added a test for this.

**Comment by lorisleiva** [comment]
### Merge activity

* **Apr 2, 6:48 PM EDT**: @lorisleiva started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/2397).
